pkg/agent/agent.go

```diff
@@ -769,6 +780,51 @@ func (i *Initializer) initializeIPSec() error {
 	return nil
 }
 
+// initializeWireguard checks if preconditions are met for using WireGuard and initialize WireGuard client or clean up.
+func (i *Initializer) initializeWireGuard() error {
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		return err
+	}
+	node, err := i.client.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Failed to get node from K8s with name %s: %w", nodeName, err)
+		return err
+	}
+
+	ipAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		return fmt.Errorf("failed to obtain local IP address from K8s: %w", err)
+	}
+	_, localIntf, err := getIPNetDeviceFromIP(ipAddr)
+	if err != nil {
+		return err
+	}
+	mtu := localIntf.MTU - config.WireGuardOverhead
+	//create and initialize wireguard client
+	v4Enabled := config.IsIPv4Enabled(i.nodeConfig, i.networkConfig.TrafficEncapMode)
+	v6Enabled := config.IsIPv6Enabled(i.nodeConfig, i.networkConfig.TrafficEncapMode)
+
+	var nodeIPv4, nodeIPv6 net.IP
+	nodeIPv4 = i.nodeConfig.NodeIPAddr.IP.To4()
+
+	if nodeIPv4 == nil {
+		nodeIPv6 = i.nodeConfig.NodeIPAddr.IP.To16()
+	}
+
+	wgClient, err := wireguard.New(nodeName, i.client, mtu, nodeIPv4, nodeIPv6,
+		v4Enabled, v6Enabled, i.networkConfig.WireGuardPort)
+	if err != nil {
+		return err
+	}
+	i.wireguardClient = wgClient
+	// clean up routes
+	if !i.networkConfig.EnableWireGuard {
+		return i.wireguardClient.Cleanup()
```

Comment:
Does any other switchable features do the cleanup? e.g. ipsec?
Asking because if we do this for other features, it basically means every module should initialize and do cleanup even if they are not enabled. if we don't do this for other features, it seems not necessary to do it for this one too. Not sure if even needs the kernel module loaded to do the cleanup.

Reply:
AFAIK for IPsec, we clean up stale OVS tunnel ports after IPsec is disabled. We do not need to touch the routes/rules on the node host network as WireGuard does. I wonder whether it is safe to leave the route and rule after enabling and disabling WireGuard feature. As you mentioned that the cleanup is not a completed cleanup as it is not necessary to remove the loaded kernel module.